File: templates/python/code-carl/ch01_array/t4_lc0977_sorted_squares.py
"""LeetCode 977. Squares of a Sorted Array
https://leetcode.cn/problems/squares-of-a-sorted-array/
"""

from typing import List


class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        pass


if __name__ == "__main__":
    solution = Solution()
    assert solution.sortedSquares([-4, -1, 0, 3, 10]) == [0, 1, 9, 16, 100]
    assert solution.sortedSquares([-7, -3, 2, 3, 11]) == [4, 9, 9, 49, 121]
    print("All tests passed.")

File: solutions/java/code-carl/ch01_array/T4_LC0977_SortedSquares.java
/**
 * Practice template for:
 * - LeetCode 977. Squares of a Sorted Array
 *   https://leetcode.cn/problems/squares-of-a-sorted-array/
 */
public class T4_LC0977_SortedSquares {
    public int[] sortedSquares(int[] nums) {
        int left = 0, right = nums.length - 1;
        int[] res = new int[nums.length];
        int i = nums.length - 1;
        while (left <= right) {
            int m1 = nums[left] * nums[left];
            int m2 = nums[right] * nums[right];
            if (m1 > m2) {
                res[i--] = m1;
                left++;
            } else {
                res[i--] = m2;
                right--;
            }
        }
        return res;
    }
}
